synthesizer/process/src/stack/register_types/initialize.rs

```diff
@@ -64,64 +67,142 @@ impl<N: Network> RegisterTypes<N> {
+                num_futures == 1,
+                "Function '{}' must contain exactly one 'future' output, found {num_futures}",
+                function.name()
+            );
             ensure!(
-                command.operands().len() <= N::MAX_INPUTS,
-                "Function '{}' has too many finalize operands",
+                match function.outputs().last().map(|output| output.value_type()) {
+                    Some(ValueType::Future(locator)) =>
+                        locator.program_id() == stack.program_id() && locator.resource() == function.name(),
+                    _ => false,
+                },
+                "The last output of function '{}' must be a future associated with itself",
                 function.name()
             );
+        } else {
+            ensure!(
+                num_futures == 0,
+                "Function '{}' must not contain any 'future' outputs, found {num_futures}",
+                function.name()
+            );
+        }
 
-            // Check the type of each finalize operand.
-            for operand in command.operands() {
-                // Retrieve the register type from the operand.
-                let register_type = register_types.get_type_from_operand(stack, operand)?;
-                // Ensure the register type is a literal or a struct.
-                // See `Stack::execute_function()` for the same set of checks.
-                match register_type {
-                    RegisterType::Plaintext(PlaintextType::Literal(..)) => (),
-                    RegisterType::Plaintext(PlaintextType::Struct(..)) => (),
-                    RegisterType::Plaintext(PlaintextType::Array(..)) => (),
-                    RegisterType::Record(..) => {
-                        bail!(
-                            "'{}/{}' attempts to pass a 'record' into 'finalize'",
-                            stack.program_id(),
-                            function.name()
-                        );
-                    }
-                    RegisterType::ExternalRecord(..) => {
-                        bail!(
-                            "'{}/{}' attempts to pass an 'external record' into 'finalize'",
-                            stack.program_id(),
-                            function.name()
-                        );
-                    }
+        /* Additional checks. */
+        // - All futures produces before the `async` call must be consumed by the `async` call, in the order in which they were produced.
+
+        // Get all registers containing futures.
+        let mut future_registers = register_types
+            .destinations
+            .iter()
+            .filter_map(|(_, register_type)| match register_type {
+                RegisterType::Future(locator) => Some(*locator),
+                _ => None,
+            })
+            .collect::<Vec<_>>();
+
+        // Remove the last locator, since this is the future created by the `async` call.
+        future_registers.pop();
```

Comment:
the allocation to a `Vec` can be avoided here (by leaving it in the iterator form), as well as with `async_future_operands`; the final check could instead be `async_future_operands.zip(future_registers).all(|(afo, fr)| afo == fr)`; in order to make sure the lengths match, it can be followed with `ensure!(future_registers.next().is_none(), ...)`

Reply:
How do you get around `zip` takes ownership of `future_registers`?

Reply:
via `Iterator::by_ref`